subsys/bluetooth/mesh/adv_ext.c

```diff
@@ -172,122 +220,175 @@ static void send_pending_adv(struct k_work *work)
 		}
 	}
 
-	if (!IS_ENABLED(CONFIG_BT_MESH_GATT_SERVER)) {
+	if (IS_ENABLED(CONFIG_BT_MESH_GATT_SERVER) &&
+	    (adv->tag & BT_MESH_PROXY_ADV) &&
+	    !bt_mesh_adv_gatt_send()) {
+		atomic_set_bit(adv->flags, ADV_FLAG_PROXY);
 		return;
 	}
-
-	/* No more pending buffers */
-	if (bt_mesh_is_provisioned()) {
-		if (IS_ENABLED(CONFIG_BT_MESH_GATT_PROXY)) {
-			err = bt_mesh_proxy_adv_start();
-			BT_DBG("Proxy Advertising");
-		}
-	} else if (IS_ENABLED(CONFIG_BT_MESH_PB_GATT)) {
-		err = bt_mesh_pb_gatt_adv_start();
-		BT_DBG("PB-GATT Advertising");
-	}
-
-	if (!err) {
-		atomic_set_bit(adv.flags, ADV_FLAG_PROXY);
-	}
 }
 
-static void schedule_send(void)
+static bool schedule_send(struct bt_mesh_ext_adv *adv)
 {
-	uint64_t timestamp = adv.timestamp;
+	uint64_t timestamp;
 	int64_t delta;
 
-	if (atomic_test_and_clear_bit(adv.flags, ADV_FLAG_PROXY)) {
-		bt_le_ext_adv_stop(adv.instance);
-		atomic_clear_bit(adv.flags, ADV_FLAG_ACTIVE);
+	if (!adv) {
+		return false;
 	}
 
-	if (atomic_test_bit(adv.flags, ADV_FLAG_ACTIVE) ||
-	    atomic_test_and_set_bit(adv.flags, ADV_FLAG_SCHEDULED)) {
-		return;
+	timestamp = adv->timestamp;
+
+	if (atomic_test_and_clear_bit(adv->flags, ADV_FLAG_PROXY)) {
+		(void)bt_le_ext_adv_stop(adv->instance);
+		atomic_clear_bit(adv->flags, ADV_FLAG_ACTIVE);
+	}
+
+	if (atomic_test_bit(adv->flags, ADV_FLAG_ACTIVE) ||
+	    atomic_test_and_set_bit(adv->flags, ADV_FLAG_SCHEDULED)) {
+		return false;
 	}
 
 	/* The controller will send the next advertisement immediately.
 	 * Introduce a delay here to avoid sending the next mesh packet closer
 	 * to the previous packet than what's permitted by the specification.
 	 */
 	delta = k_uptime_delta(&timestamp);
-	k_work_reschedule(&adv.work, K_MSEC(ADV_INT_FAST_MS - delta));
+	k_work_reschedule(&adv->work, K_MSEC(ADV_INT_FAST_MS - delta));
+
+	return true;
 }
 
-void bt_mesh_adv_update(void)
+void bt_mesh_adv_gatt_update(void)
 {
-	BT_DBG("");
+	(void)schedule_send(gatt_adv_get());
+}
 
-	schedule_send();
+void bt_mesh_adv_buf_local_ready(void)
+{
+	(void)schedule_send(&adv_main);
 }
 
-void bt_mesh_adv_buf_ready(void)
+void bt_mesh_adv_buf_relay_ready(void)
 {
-	schedule_send();
+	int i = 0;
+	struct bt_mesh_ext_adv *adv = relay_adv_get();
+
+	do {
+		if (schedule_send(adv++)) {
+			return;
+		}
+	} while (CONFIG_BT_MESH_RELAY_ADV_SETS > ++i);
```

Comment:
Could you just convert this to the more usual for-loop?
```
for (int i = 0; i < CONFIG_BT_MESH_RELAY_ADV_SETS; i++) {
        if (schedule_send(&adv[i])) {
                return;
        }
}
```

I've used `&adv[i]` there instead of `adv++` since iterating a pointer you got from some function return looks dangerous. It's good to make it clear that you're treating this pointer as an array whose valid indices follow the same bounds as the `i` variable.